spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -156,17 +180,26 @@ public void setExecutor(Executor executor) {
 	public Lock obtain(Object lockKey) {
 		Assert.isInstanceOf(String.class, lockKey);
 		String path = (String) lockKey;
-		return this.locks.computeIfAbsent(path, RedisLock::new);
+		synchronized (locks){
+			return this.locks.computeIfAbsent(path, RedisLock::new);
+		}
 	}
 
 	@Override
 	public void expireUnusedOlderThan(long age) {
 		long now = System.currentTimeMillis();
-		this.locks.entrySet()
-				.removeIf((entry) -> {
-					RedisLock lock = entry.getValue();
-					return now - lock.getLockedAt() > age && !lock.isAcquiredInThisProcess();
-				});
+		synchronized (locks) {
+			Iterator<Entry<String, RedisLock>> iterator = locks.entrySet().iterator();
+			while (iterator.hasNext()) {
+				Entry<String, RedisLock> entry = iterator.next();
+				RedisLock lock = entry.getValue();
+				if (now - lock.getLockedAt() > age && !lock.isAcquiredInThisProcess()) {
+					iterator.remove();
+				} else {
+					break;
```

Comment:
> I don't think this is correct assumption to stop iterating the cache because we don't match that `getLockedAt()` any more. The cache is only for instances to not overhead with them for the same key, but any obtained `Lock` can be locked later in the logic at arbitrary time. Therefore the oldest entry in the cache might be the most fresh with its `lock()` call. So, early exit from such a loop is misleading and may leave really old not locked entries in the cache.

Thanks for the review.

Fixed to be sorted when lock() is called.
For removeEldestEntry() and expireUnusedOlderThan()

Is it ok to change like this?

Reply:
It is not. 
See my comment on the `obtainLock()` below.